internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -408,7 +410,10 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 		// Specifically it will merge rather than replace owner references (e.g.
 		// for Usages), and will fail if we try to add a controller reference to
 		// a resource that already has a different one.
-		if err := c.client.Patch(ctx, cd.Resource, client.Apply, client.ForceOwnership, client.FieldOwner(FieldOwnerComposed)); err != nil {
+		// NOTE(phisco): We need to set a field owner unique for each XR here,
+		// this prevents multiple XRs composing the same resource to be
+		// continuously alternated as controllers.
+		if err := c.client.Patch(ctx, cd.Resource, client.Apply, client.ForceOwnership, client.FieldOwner(xr.GetName())); err != nil {
```

Comment:
Would it make sense to prefix the XR name with `FieldOwnerComposed`? Assuming there's no length issues it might make identifying field owners easier.

Reply:
@phisco FYI this is the only thing blocking me approving.

Reply:
Sorry missed it, I'll double-check there is no length limit and add that if not.

Reply:
should be at most 128 characters:
<img width="1299" alt="image" src="https://github.com/crossplane/crossplane/assets/5697904/281d9039-bfcc-4016-ac94-63987a9c3441">

Reply:
which is much shorter than the 253 allowed characters for names 🤔 we should either crop it or hash it actually.

Reply:
Could we use the XR's UID? `FieldOwnerComposed-<36 chars>` should fit.

The only issue I can think of would be if we wanted to backup and restore the XR and all of its composed resources. The restored XR would have a new UID.

Reply:
I'm not sure whether we would preserve managed fields restoring 🤔 however, what we want to avoid at all costs is collisions, ownership is already clearly stated in the ownerRefs, so the prefix + UID or hash could be both fine, but the hash would be stable w.r.t. the name, so there would be no risk at all of issues at restore time, while the UID could potentially cause some.

Reply:
Yeah my guess is that UID might be a non-starter. CC resident backup-and-restore expert @turkenh.

Hash of the name SGTM, ideally with some human readable prefix.

Reply:
> I'm not sure whether we would preserve managed fields restoring

No, we don't. UIDs change after backup/restore. 
+1 for hashing the name.

Reply:
🤔 @negz @turkenh, should I also include the GVK in the hash? I feel it could cause some issues in case of version bumps of the XR, so the version should not definitely be hashed, but what about the group and kind? is there any possible scenario where a user is willingly changing the XR of a set of composed resources altogether, except for the name?

Reply:
have a look at 
https://github.com/crossplane/crossplane/pull/5236/commits/687bf3eb3a4d8a293e8acb7abbfd78f7c5b8487f

Reply:
I _think_ that approach SGTM.

The only thing I'm wondering is whether we're thinking about field owners and controller references the right way.

[The SSA docs](https://kubernetes.io/docs/reference/kubernetes-api/common-parameters/common-parameters/#fieldManager) say the field manager (aka owner) should be "the actor or entity that is making the changes". I can't find clear guidance, but for a child resource produced by a controller I'd intuitively consider the actor to be the controller doing the reconciling, not the parent resource being reconciled.

XR controllers are GVK-scoped. One XR controller owns and reconciles all XRs of a particular GVK. You're right that the version could change over time, so let's say they're GK scoped. I'd therefore expect that the field manager would be `apiextensions.crossplane.io/composite/example.org/Widget` - without a name. It certainly doesn't seem like the `fieldManager` field was _designed_ to fit enough information to uniquely identify a resource, unless they were expecting us to use a UID?

[The controller reference design doc](https://github.com/kubernetes/design-proposals-archive/blob/main/api-machinery/controller-ref.md) seems to support this expectation. It states that the main goal of a controller reference is to solve the problem of controllers that fight over controlled objects. It uses two _different_ controllers - ReplicaSet and ReplicationController - to illustrate this example.

On the other hand, a controller reference is a reference to a specific resource, not just a controller. We've certainly always used controller references to ensure one parent resource does not accidentally adopt a child that belongs to another parent of the same type, with the same controller.

I also can't see how it would be possible to write a truly [reconstructive controller](https://kubernetes.io/blog/2022/10/20/advanced-server-side-apply/#reconstructive-controllers) without uniquely identifying the parent resource as the field manager. If you use only the controller as the field manager, the controller would always need to read a composed resource and check its controller reference before applying desired state. Otherwise SSA would allow two XRs of the same kind (reconciled by the same controller) to fight over control for the same composed resource.

Reply:
Yes, it is odd indeed. Looks like the assumption is that a specific controller is only going to reconcile a resource for a single reason, hence a single name per controller is enough, while in our case the same XR controller could be reconciling the same composed resource for multiple XRs, hence we need a unique name for each XR to allow SSA to properly prevent hijackings.